sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/SupportsOverwrite.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.writer;
+
+import org.apache.spark.sql.sources.Filter;
+
+/**
+ * Write builder trait for tables that support overwrite by filter.
+ * <p>
+ * Overwriting data by filter will delete any data that matches the filter and replace it with data
+ * that is committed in the write.
+ */
+public interface SupportsOverwrite extends WriteBuilder {
```

Comment:
ah i see, make senses. So this `SupportsOverwrite` is mostly for partitioned tables, though it doesn't mention partition at all. I think the semantic can also apply to non-partitioned tables but that will be very hard to implement.

Reply:
BTW for a normal INSERT OVERWRITE, which needs to truncate the entire table, the `filters` will be a single true literal?

Reply:
> I think the semantic can also apply to non-partitioned tables but that will be very hard to implement.

Not necessarily. JDBC sources can implement this fairly easily, and those that support transactions can make it an atomic operation. There are also strategies that can work for unpartitioned tables, like deleting data files using min/max ranges show all rows are matched by a filter.

> BTW for a normal INSERT OVERWRITE, which needs to truncate the entire table, the filters will be a single true literal?

I think that truncate is slightly different. Because it is fairly easy to support truncate, but not overwrite by expression, I think that truncate should be a separate operation in the v2 API. I would make `SupportsOverwrite` implement `SupportsTruncate` with a default that calls overwrite with `true` like you suggest, but I think we will need to add a `true` filter.

Also, what do you mean by "normal INSERT OVERWRITE"? What operation is that? Right now, `INSERT OVERWRITE` is effectively dynamic partition overwrite. Unpartitioned tables are truncated because they have just one "partition". Do you agree with that summary?

Reply:
> Unpartitioned tables are truncated because they have just one "partition"

If we go with this definition, do you mean data sources that do not support partition must implement `SupportsDynamicOverwrite` to support `INSERT OVERWRITE`?

Reply:
@cloud-fan, I've added `SupportsTruncate` to address this. Now tables can support truncation to implement `INSERT OVERWRITE` that replaces all data in the table.

I also ran tests to find out how file system data sources behave with overwrite mode in `DataFrameWriter`. That showed that tables are always truncated and the value of `spark.sql.sources.partitionOverwriteMode` is always ignored.